[WIP] CDS dashboard tutorial (#52)
nebari-dev/nebari: docs/docs/tutorials/creating-cds-dashboard.md

## Deploying with CDS dashboards

Now we'll use CDS Dashboards and publish this notebook. You can visit [CDS Dashboards docs](https://cdsdashboards.readthedocs.io/en/stable/) for more details on this great package.

To begin, click on the top left tab navigate to `File` > `Hub Control Panel` > `Dashboards`. Click on the button `New Dashboard`.

### Steps to create new dashboard and publish

- Add a dashboard name
- Add a short description
- Customise user permission (optional)
- Select the file source, for this tutorial we will select `Jupyter Tree`
- Details section
  - Select the framework of your choice, here we will select `panel`
  - Select the conda environment, make sure it is same as Jupyter notebook environment
  - In the `relative path` box, copy your notebook's path (example: `demo-dashboards/tutorial/insights.ipynb`)
  - Click on the save button
  - Select the instance size and save again, a message should appear stating: **The dashboard is starting up**

![CDS dashboard configuration](/img/cds_details.png)

![The dashboard is starting up](/img/dashboard_starting_up.png)

### More on instance selection and publishing dashboard

> When we publish our dashboard, compute and storage resources are allocated. The selection of right set of resources

Reviewer: ```suggestion
When we publish our dashboard, compute and memory resources are allocated. The instance size selected
```

> would differ based on use-case. Currently, _Nebari_ offers four instance types namely `Small`, `Medium`, `Large` and `Large GPU`.

Reviewer: ```suggestion
will differ based on use-case. Currently, _Nebari_ offers two instance types namely `Small` and `Medium`. This deployment also has added two instance types added as options, `Large` and `Large GPU`.
```

If required the instances could be scaled up/down manually (auto-scaling is not supported).

![Instance selection](/img/select_instance.png)

> Once the instance is selected, the server is spun. If there are no errors encountered during this process, we will

Reviewer: ```suggestion
Once the instance is selected, the server is spun up. If there are no errors encountered during this process, we will
```

---
id: creating-cds-dashboard
title: How to create a containDS dashboard on nebari?
---

# What is a dashboard?

Data provides us with insights, a dashboard stitches these insights together into a meaningful story.

Different tools and technologies could be used to create and deploy a dashboard. In this tutorial
we will be building a simple dashboard on _nebari_ which could be shared with other users and groups via
[ContainDS Dashboards](https://cdsdashboards.readthedocs.io/en/stable/).

Currently, this dash-boarding solution supports `Panel`, `Bokeh`, `Voila`, `Streamlit`, and `Plotly`.

# Let's get our hands dirty

- Create a notebook in your jupyterlab environment on nebari with a name of your choice

> - Select an environment from the dropdown (eg: `dashboard`)

Author: Was referring to kernel dropdown, I shall update the wordings.

           site="About 🌳",
           title="Specie info and more",
           main=[combine_plot]
           )
dashboard.servable()
```

# Step-by-step guide for building a CDS dashboard

Let's use CDS and publish this notebook, visit [CDS docs](https://cdsdashboards.readthedocs.io/en/stable/) for more details.
Click on the top left tab navigate to `file` > `Hub Control Panel` > `Dashboards`. Click on the button `New Dashboard`.

> Create new dashboard steps

- Add a dashboard name (mandatory)
- Add a short description (mandatory)
- Customise user permission (optional)
- Select the file source, for this tutorial we will select `Jupyter Tree` (mandatory)

> Details section

- Select the framework of your choice, here we will select `voila` (mandatory)
- Select the conda env, make sure it is same as jupyter notebook environment (mandatory)
- In the relative path copy your notebook's path (example: `demo-dashboards/tutorial/insights.ipynb`) (mandatory)
- Click on the save button
- Select the instance size and save, a message should appear stating -> **The dashboard is starting up**

![CDS dashboard configuration](/img/cds_details.png)

# Conclusion

> We have a working dashboard now 🎉 sharing the link is super simple, copy-paste the URL and voilà.

Reviewer 2: ```suggestion
We have a working dashboard now 🎉   
All that is left to do is to copy the URL and send it to your colleagues!
```

> It is fast and customisable to your need.

Reviewer 2: ```suggestion
```

---
id: creating-cds-dashboard
title: How to create a ContainDS dashboard on Nebari
---

> 

Reviewer 2: ```suggestion

# Create a dashboard with Panel and Deploy it with CDS Dashboards

```

> # What is a dashboard?

Reviewer 2: ```suggestion
## What is a dashboard?
```

Analyzing data provides us with insights. A dashboard stitches these insights together into a meaningful story.

Different tools and technologies can be used to create and deploy a dashboard. In this tutorial
we will be building a simple dashboard on _Nebari_ which can be shared with other users and groups via
[ContainDS Dashboards](https://cdsdashboards.readthedocs.io/en/stable/) (CDS Dashboards).

Currently, CDS Dashboards supports `Panel`, `Bokeh`, `Voila`, `Streamlit`, and `Plotly`.
We will be using `Panel` in this tutorial to create a dashboard.

> # Let's get our hands dirty

Reviewer 2: ```suggestion
## Creating the dashboard

*Let's get our hands dirty*

```

- Create a notebook in your JupyterLab environment on Nebari with a name of your choice
- Select an environment from the select kernel dropdown (located on the top right: e.g. `dashboard`)
- Copy the code below to your notebook.

> We can now execute the notebook and view our Panel dashboard as output right inside the notebook.
> Making it super easy to rapidly prototype new dashboards. Once we are happy with what we've created,
> we can move on to deploying it with CDS Dashboards.

Reviewer 2: ```suggestion
```

> The code below dives into creating a dashboard from insights.

Reviewer 2: ```suggestion
```

> ##### Creating insights using `holoviews`, `panel` and `bokeh` as backend

Reviewer 2: ```suggestion
Our dashboard will be creating insights using `holoviews` and `panel` with `bokeh` as backend.
```

```python

> # necessary imports

Reviewer 2: ```suggestion
```

> import numpy as np

Reviewer 2: ```suggestion
```

import pandas as pd
import holoviews as hv
from bokeh.models import HoverTool

> from holoviews import opts

Reviewer 2: ```suggestion
```

import panel as pn
hv.extension('bokeh')

> 

Reviewer 2: are you sure you don't need a `pn.extension()` here? I think leaving it off means this dashboard will work in some cases, but not in others.

Author: For the last few EDA/dashboards, I didn't include `pn.extension()` and didn't face any issues. 
Though I have updated the code for the tutorial (to eliminate any unforeseen errors)

# creating a sample dataset
data_trees = { 'species_name': ['live oak', 'pecan', 'bur oak', 'cedar elm'],
               'avg_diameter_inch': [20, 30, 40, 35]
             }

df = pd.DataFrame(data_trees)

# simple curve/line and bar plots
plot_bar = hv.Bars(df, 'species_name', 'avg_diameter_inch')
plot_curve = hv.Curve(df)

# creating hover tooltip
hover = HoverTool(tooltips=[("avg diameter", "@avg_diameter_inch"),
                            ("species", "@species_name")])
# plot customization
combine_plot = plot_bar.opts(tools=[hover]) + plot_curve.opts(line_dash='dashed')

# creating simple dashboard using panel
dashboard = pn.template.BootstrapTemplate(
           site="About 🌳",
           title="Species and more",
           main=[combine_plot]
           )
dashboard.servable()
```

> 

Reviewer 2: ```suggestion
We can now execute the notebook and view our Panel dashboard as output right inside the notebook. This feature Panel makes it very easy to rapidly prototype new dashboards. Once we are happy with what we've created, we can move on to deploying it with CDS Dashboards.
```

> # Step-by-step guide for building a CDS dashboard

Reviewer 2: ```suggestion
## Deploying with CDS dashboards
```

> Let's use CDS and publish this notebook, visit [CDS docs](https://cdsdashboards.readthedocs.io/en/stable/) for more details.

Reviewer 2: ```suggestion
Now we'll use CDS Dashboards and publish this notebook. You can visit [CDS Dashboards docs](https://cdsdashboards.readthedocs.io/en/stable/) for more details on this great package.
```

> Click on the top left tab navigate to `file` > `Hub Control Panel` > `Dashboards`. Click on the button `New Dashboard`.

Reviewer 2: ```suggestion

To begin, click on the top left tab navigate to `File` > `Hub Control Panel` > `Dashboards`. Click on the button `New Dashboard`.
```

> > Steps to create new dashboard and publish

Reviewer 2: ```suggestion
### Steps to create new dashboard and publish
```

> - Add a dashboard name (mandatory)
> - Add a short description (mandatory)
> - Customise user permission (optional)
> - Select the file source, for this tutorial we will select `Jupyter Tree` (mandatory)

Reviewer 2: ```suggestion
- Add a dashboard name
- Add a short description
- Customise user permission (optional)
- Select the file source, for this tutorial we will select `Jupyter Tree`
```

> 
> **Details section**
> 

Reviewer 2: ```suggestion
- Details section
```

> - Select the framework of your choice, here we will select `panel` (mandatory)
> - Select the conda environment, make sure it is same as Jupyter notebook environment (mandatory)
> - In the `relative path` box, copy your notebook's path (example: `demo-dashboards/tutorial/insights.ipynb`) (mandatory)
> - Click on the save button
> - Select the instance size and save, a message should appear stating :- **The dashboard is starting up**

Reviewer 2: ```suggestion
  - Select the framework of your choice, here we will select `panel`
  - Select the conda environment, make sure it is same as Jupyter notebook environment 
  - In the `relative path` box, copy your notebook's path (example: `demo-dashboards/tutorial/insights.ipynb`) 
  - Click on the save button
  - Select the instance size and save, a message should appear stating: **The dashboard is starting up**
```

Reviewer 2: We need more details on what happens after you click the save button. I'd like to see a png of the instances I can choose from. Tell me what I need to think about when choosing one. 

I'd like to see a png of the dashboard server starting up. 

Then I need details of how to get to that server. So when I first deploy it, i should be redirected to my running deployment. After that, if I need to go back and find it, I go to the `Dashboards` page and see a list. Tell me about that. 

Add a note warning people that if you leave a dashboard running, its using resources. So for example, if you're on a cloud deployment, it may be incurring high cost. 

Add information on how to stop a deployment, and how to delete a dashboard. 

Add information on where to find your deployments and dashboards others have deployed. (with pngs as appropriate).
